cmd/tether/attach.go

```diff
@@ -445,52 +510,66 @@ func (t *attachServerSSH) run() error {
 			log.Debugf("reader/writers bound for channel for %s", sessionid)
 
 			go t.channelMux(requests, session, detach)
-
-			if session.RunBlock && session.ClearToLaunch != nil && session.Started != "true" {
-				log.Debugf("Unblocking the launch of %s", sessionid)
-				// make sure that portlayer received the container id back
-				session.ClearToLaunch <- <-t.askedAndAnswered
-				log.Debugf("Unblocked the launch of %s", sessionid)
-			}
 		}
 		log.Info("Incoming attach channel closed")
 	}
 	return nil
 }
 
-func (t *attachServerSSH) globalMux(reqchan <-chan *ssh.Request) {
-	defer trace.End(trace.Begin("attach server global request handler"))
+func (t *attachServerSSH) sessions(all bool) []string {
+	defer trace.End(trace.Begin(""))
 
-	// to make sure we close the channel once
-	var once sync.Once
+	var keys []string
+
+	// this iterates the local copies of the sessions maps
+	// so we don't need to care whether they're initialized or not
+	// as extension reload comes after that point
+
+	// whether include sessions or not
+	if all {
+		for k, v := range t.config.Sessions {
+			if v.Active && v.StopTime == 0 {
+				keys = append(keys, k)
+			}
+		}
+	}
+
+	for k, v := range t.config.Execs {
+		// skip those that have had launch errors
+		if v.Active && v.StopTime == 0 && (v.Started == "" || v.Started == "true") {
+			keys = append(keys, k)
+		}
+	}
+
+	log.Debugf("Returning %d keys", len(keys))
+	return keys
+}
 
-	// ContainersReq will close this channel
-	t.askedAndAnswered = make(chan struct{})
+func (t *attachServerSSH) globalMux(reqchan <-chan *ssh.Request, cleanup func()) {
+	defer trace.End(trace.Begin("attach server global request handler"))
 
+	// cleanup function passed by the caller
+	defer cleanup()
+
+	// for the actions after we process the request
+	var pendingFn func()
 	for req := range reqchan {
-		var pendingFn func()
 		var payload []byte
 		ok := true
 
 		log.Infof("received global request type %v", req.Type)
 
 		switch req.Type {
 		case msgs.ContainersReq:
-			keys := make([]string, len(t.config.Sessions))
-			i := 0
-			for k := range t.config.Sessions {
-				keys[i] = k
-				i++
+			msg := msgs.ContainersMsg{
+				IDs: t.sessions(true),
 			}
-			msg := msgs.ContainersMsg{IDs: keys}
 			payload = msg.Marshal()
-
-			// unblock ^ (above)
-			pendingFn = func() {
-				once.Do(func() {
-					close(t.askedAndAnswered)
-				})
+		case msgs.VersionReq:
+			msg := msgs.VersionMsg{
+				Version: feature.MaxPluginVersion - 1,
```

Comment:
This will turn into an int during the compilation and I believe that is what we want. We are trying to learn the tether binary's version not the guestinfo value